datastore/concurrentds.nim

```diff
@@ -0,0 +1,50 @@
+import pkg/chronos
+import pkg/questionable
+import pkg/questionable/results
+import pkg/upraises
+
+import ./key
+import ./query
+import ./types
+import ./datastore
+
+export key, query, types, datastore
+
+push: {.upraises: [].}
+
+type
+  Function*[T, U] = proc(value: T): U {.upraises: [], gcsafe, closure.}
+  Modify* = Function[?seq[byte], ?seq[byte]]
+  ModifyAsync* = Function[?seq[byte], Future[?seq[byte]]]
+
+method modify*(self: ConcurrentDatastore, key: Key, fn: Modify): Future[?!void] {.base, locks: "unknown".} =
```

Comment:
If you have a sync callback, there is no reason to make the top level method `async`, i.e. return a `Future`.

Reply:
This method from `Datastore` interface doesn't have any callback at all and it returns future:

`method put*(self: Datastore, key: Key, data: seq[byte]): Future[?!void]`

Reply:
True, and this is one thing I want to change, there should be a sync and an async interface, (but this ins't as straightforward), in which case those methods won't return a `Future`. 

The problem with what you're doing is that you are adding a blocking callback to a non-blocking method call, when executing the callback, it's going to block the "non-blocking" call.

Reply:
Let's be specific. By blocking callback you mean a function containing call that will block a thread? Like [std/os sleep](https://nim-lang.org/docs/os.html#sleep%2Cint) for example? If so, what prevents users from using such call in an async function?

Reply:
Nothing, but why return a `Future` at all if the callback signature is already _blocking_ and there is no way of calling it in a non blocking way - you see the discrepancy? 

All I'm  saying is that, if you "promising" that a method/proc is asynchronous, it doesn't make sense to pass in a synchronous primitive, so either make the callback async or remove the `Future` return type. You already have an async version of this method, so the logical step would be to have a sync version of it, simply do not return a `Future` from the sync version.

Reply:
>the callback signature is already blocking

A signature cannot be blocking. An implementation can be blocking. There's nothing in this signature that says that the implementation has to be blocking. It can be a simple pure function (and that's the intention). Example: adding +1 to a number (like increasing a ref count) - this is a pure and non-blocking operation.

The method signature returns Future because a store that will implement it will likely do an IO operation underneath, similarly as `put` is doing.

Reply:
>a blocking callback

>a synchronous callback

You're using term `synchronous` and `blocking` interchangeably, as if they would be the same thing - which is not the case.

This call to `+` is synchronous and non blocking `let x = 1 + 1`.
This call to `sleep` is synchronous and blocking `sleep(100)`.

>anyone can simply assume that since it still returns a Future, it would be safe to do blocking or long running operations in the callback.

It's never-ever safe to do any blocking or long running calls in the code if this code is going to be be executed on a main compute thread (the event loop thread). And that's it. If anyone assumes that it's safe to use a blocking call in the callback, then well - that's simply a wrong assumption.

Anyway, if you're still unconvinced I'm going to remove these methods, just to get a review passed.

Reply:
`synchronous` and `blocking` are interchangeable in this context, both long running computations and blocking IO will have the same effect on the event loop.

> If anyone assumes that it's safe to use a blocking call in the callback, then well - that's simply a wrong assumption.

That isn't the case tho, any long running computation can be broken down into smaller pieces, yielding control back to the main event loop, for example you can use `idleAsync` or `sleepAsync`, but this is only possible in an async callback, i.e. one that returns a continuation.

Reply:
But I think the finer point of my argument is being missed thought the entire exchange - it is simply wrong to mix incompatible execution policies, this brings confusion and ultimately leads to a bad API.

Reply:
Ok, I understand. Methods accepting non-async callbacks got removed in latest changes.